_drafts/best-of-binutils.md

```diff
@@ -0,0 +1,1011 @@
+
+```
+$ readelf -S <ELF>
+  ...
+  [ 8] .debug_info       PROGBITS        00000000 0200f4 00980b 00      0   0  1
+  [ 9] .debug_abbrev     PROGBITS        00000000 0298ff 001adf 00      0   0  1
+  [10] .debug_loc        PROGBITS        00000000 02b3de 003bf7 00      0   0  1
+  [11] .debug_aranges    PROGBITS        00000000 02efd5 0004f0 00      0   0  1
+  [12] .debug_ranges     PROGBITS        00000000 02f4c5 0007d8 00      0   0  1
+  [13] .debug_macro      PROGBITS        00000000 02fc9d 0031dd 00      0   0  1
+  [14] .debug_line       PROGBITS        00000000 032e7a 005447 00      0   0  1
+  [15] .debug_str        PROGBITS        00000000 0382c1 00dedd 01  MS  0   0  1
+```
+
+### stabs, COFF, XCOFF
+
+While rather infrequent, if you are working with custom compilers (such as
+those shipped with some DSPs), you may see binary / debug information emitted in alternative formats such as
+stabs, COFF, or XCOFF. These are all legacy formats that have largely been superseded by ELF +
+DWARF. I view it as a red-flag in the evaluation process if a vendors compiler emits information in
+these formats. At this point, a lot of binary inspection tools have deprecated support
+for these formats. This makes it harder to debug and workaround issues on chips using these formats.
+
+## Examples
+
+All the example code used in the sections below can be found in it's entirety
+[on Github](https://github.com/memfault/interrupt/tree/master/example/best-of-binutils). If you
+would like to follow along and build any of the sources yourself you can find more information there.
+
+## GNU Binutils
+
+GNU Binutils is a collection of supporting tools that work alongside GNU GCC to inspect and edit binaries.
+We'll go into details below but most of the tools operate on ELFs. On linux and for embedded
+devices both individual object files (`.o`) and the final linked output (the ELF) are in this format.
+
+### Compiling Binutils From Source
+
+Sometimes it can be fun (or useful) to compile the latest GNU Binutil to get a preview of
+unreleased features or customize the supported architecture targets that are supported:
+
+```
+$ git clone git://sourceware.org/git/binutils-gdb.git
+$ git log --oneline -1
+3349112e38 elf: Remove zero-sized relocation section from section group
+$ ./configure --prefix=/usr/local/binutils_latest/arm-none-eabi --target=arm-none-eabi
+--with-python --disable-ld
+$ make
+$ # Wait 10-20 minutes ...
+$ ./binutils/size --version
+GNU size (GNU Binutils) 2.34.50.20200404
+Copyright (C) 2020 Free Software Foundation, Inc.
+This program is free software; you may redistribute it under the terms of
+the GNU General Public License version 3 or (at your option) any later version.
+This program has absolutely no warranty.
+```
+
+### Compiling Binutils For Multiple Targets
+
+A typical GNU Binutil release will only include select targets. For example, the official GNU ARM
+toolchain[^4] includes:
```

Comment:
I finally learned to use keywords instead of numbers (sincee I find them difficult to keep track of). 

So, if you'd like to you can do:


```
toolchain[^toolchain] includes:


[^toolchain]: <link
```